### DISCUSSION TITLE
RISC-V Support

### DISCUSSION THREAD

**Discussion by archaic-archea** [LibertyOS-Development/kernel#21] [Ideas]
Hey, I've been working on OS kernels for a good little bit, and I was wondering, would working on RISC-V support would be welcome, or is it an x86-64 only OS?

**Comment by danielteberian** [comment]
It would be a welcome addition. If you can do it, I would happily add your code into the kernel.

**Comment by archaic-archea** [comment]
Dope! I'm gonna look into the kernel and what I can do to support RISC-V and I'll likely make a branch for it.

**Comment by archaic-archea** [comment]
Sadly, looking into it the OS uses the bootloader crate which is x86-64 exclusive.

**Reply by danielteberian** [reply]
One of the long-term goals of the kernel is to have no external dependencies. One of the things I have wanted to do for a while is create a bootloader for the kernel, rather than continue to use that crate.

If you would be willing to, you could try and create a bootloader. I know of a series of tutorials about writing an OS for RISC-V with Rust. I will include a link below. I have wanted to include support for RISC-V, so I would gladly assist in implementing support for it.


https://osblog.stephenmarz.com/index.html

**Reply by archaic-archea** [reply]
I hate to tell you, but currently there aren't actually that many super great tutorials on RISC-V OSDev.
The Stephen Marz one is ok, but has a lot of issues, iirc it assumes you want to start in machine mode, which I assume you don't want to as that's for firmware. That's one of the big issues at least.
That aside, I don't have much experience writing bootloaders so it'll likely be a slow thing, but I will look into it and begin architect-ing it likely.

**Reply by danielteberian** [reply]
I just found this project. Would this help?

https://lib.rs/crates/xous-kernel

**Reply by archaic-archea** [reply]
The issue isn't that I don't understand RISC-V OSDev, I have experience with RISC-V OSDev and am working on another RISC-V OS now, the issues is that Bootloaders aren't quite my cup of tea.
A bootloader I recommend is Limine, it works on x86, ARM, and RISC-V.
But if you really want a custom bootloader, I will have to await the development of that bootloader to actually work on any RISC-V port.

**Reply by danielteberian** [reply]
Would you be able to work on RISC-V support within the kernel? We may not be able to use it yet, but I don't see why it would be a bad thing to have RISC-V support ready for whenever we can get a bootloader working.
